README.md
# [Ru] Плагин «Контент - Внутритекстовая реклама»

Плагин для автоматической расстановки рекламных блоков в тексте статей.

Плагин поддерживает до 5 рекламных блоков. Код каждого блока нужно получить в личном кабинете, скопировать и вставить в настройки плагина.

Логика расстановки блоков:

* если в статье меньше 5000 символов, то реклама не вставляется. Считаются все символы: и текст, и теги с их атрибутами;
* если в статье больше 5000 символов, то на каждые 5000 символов даётся один блок;
* если в статье больше 25000 символов, то в неё вставляется пять блоков;
* первый блок вставляется после пятого параграфа;
* остальные блоки равномерно распределяются так, чтобы заполнить всё пространство до десятого параграфа с конца.

Известные проблемы:

* блоки вставляются абы как (например, могут появиться сразу после подзаголовка, а это плохо);
* правило про 10-й параграф с конца работает странно: иногда блоки появляются и в самом конце статьи, то есть после «слепой 10-параграфовой зоны» в конце. Надо смотреть, почему так.

Настройки:
* Рекламный блок #1…#5 — поля для вставки кода блоков;
* Класс блоков — CSS-класс элемента, внутри которого располагается блок;
* Отступ от начала статьи — число параграфов, начиная с первого, которые не используются для вставки блоков;
* Отступ от конца статьи — число параграфов, считая от конца статьи, куда блоки не вставляются;
* Пользовательский CSS — дополнительные стили, которые будут добавлены к странице.

# [En] Content - In-text ads Plugin

A plugin to insert ad blocks to articles text (com_content). It supports up to 5 ad blocks. You can get the code in the private area of ad network and copy-paste it to plugin settings.

Settings:
* Ad block #1…#5 — blocks code;
* Blocks class — CSS class of div's with blocks;
* Start offset, end offset — number of paragraphs from the start and from the end of the article, respectively, where ads is not ineserted;
* Custom CSS — additional styles added to the nead of the document.





Changelog
1.0.0 — March 15, 2022
Initial release.

1.1.0 - March 16, 2022
Fix a bug with number of ads on a page.
Add bottom margin to ad blocks.

1.2.0 - May 10, 2022
Add options to set blocks class, specify start and end offset, add custom CSS.
Refactor code to fit Joomla guidelines.
Add updater.


intextads.php
<?php
/**
 * @package     Intext Ads, a Joomla plugin
 *
 * @author      Pavel Syomin
 * @copyright   Copyright © Pavel Syomin, 2022 
 * @license     GNU General Public License version 3; see LICENSE.txt
 */

// No direct access to this file
defined('_JEXEC') or die();

use Joomla\CMS\Plugin\CMSPlugin;

class plgContentIntextads extends CMSPlugin
{

	protected $app;

    public function onContentPrepare($context, &$row, &$params, $page = 0)
    {   
        // Do not show ads in API mode
        if ($this->app->isClient('api'))
        {
            return;
        }
        
        // Show only in articles
        if ($context == 'com_content.article')
        {
            // Skip articles with a snippet disabling ads
            if (strpos($row->text, '{intextads:off}') !== false
                or strpos($row->text, '{intextads: off}') !== false)
            {
                return;
            }
            
            // Get ad blocks from plugin settings
            $blocks = [];
            for ($i == 1; $i <= 5; $i++)
            {
                $block = trim($this->params->get("block$i", ""));
                if ($block != '')
                {
                    $blocks[] = $block;
                }
            }
            
            if (count($blocks) == 0)
            {
                return;
            }
            
            // Place blocks in article text
            $this->_place($blocks, $row);
            
            // Add custom CSS to document head
            $this->_addCSS();
        } 
	}
    
    private function _place($blocks, &$row)
    {
        // Calculate the number of blocks to add
        $textLength = strlen($row->text);
        if ($textLength <= 25000)
        {
            $maxNBlocks = intdiv($textLength, 5000);
        }
        else
        {
            $maxNBlocks = 5;
        }
        $nBlocks = min(count($blocks), $maxNBlocks);
        
        if ($nBlocks == 0)
        {
            return;
        }
        
        // Calculate blocks frequency
        $nParagraphs = substr_count($row->text, '<p');
        $startOffset = $this->params->get('startOffset', 5);
        $endOffset = $this->params->get('endOffset', 10);
        $nParagraphsAvail = $nParagraphs - $startOffset - $endOffset;
        $blocksFreq = intdiv($nParagraphsAvail, $nBlocks);
        
        if ($nParagraphs < 5 || $nParagraphsAvail < 0)
        {
            return;
        }
        
        // Add blocks using calculated start and frequency
        $pos = strpos($row->text, '<p');
        $count = 0;
        $blocksClass = $this->params->get('blocksClass', 'intext-ads');
        while ($pos !== false)
        {
            $offset = $pos + strlen('<p');
            $pos = strpos($row->text, '<p', $offset);
            $count++;
            if ($count == $startOffset)
            {
                $block = array_shift($blocks);
                $block = "<div class=\"$blocksClass\">\n" . $block . "\n</div>\n";
                $row->text = substr_replace($row->text, $block, $pos, 0);
                $startOffset += $blocksFreq;
                $nBlocks--;
                if ($nBlocks == 0)
                {
                    break;
                }
            }
        }        
    }
    
    private function _addCSS()
    {
        $css = $this->params->get('css', '');
        $this->app->getDocument()->addStyleDeclaration($css);     
    }
}


plg_intext_ads.php
<?php

// No direct access to this file
defined('_JEXEC') or die();

use Joomla\CMS\Plugin\CMSPlugin;
use Joomla\Event\Event;
use Joomla\Event\SubscriberInterface;

class PlgContentIntextAds extends CMSPlugin
{
	/**
	 * @var    \Joomla\CMS\Application\SiteApplication
	 *
	 * @since  3.9.0
	 */
	protected $app;

	 public function onContentPrepare($context, &$row, &$params, $page = 0)
	 {
		/*
		 * Plugin code goes here.
		 * You can access parameters via $this->params
		 */
		if ($this->app->isClient('api') || $context === 'com_finder.indexer')
		{
			return;
		}
        
        if ($context !== 'com_content.article' && $view !== 'article' && $row->alias != 'golubie-ozera-aleksandrovska')
        {
            return;
        }
        
        $blocks_str = $this->params->get('blocks', '');
        if ($blocks_str == '')
        {
            return;
        }
        $blocks_str = str_replace(' ', '', $blocks_str);
        $blocks = explode(',', $blocks_str, $limit = 5);
        
        $this->_place($blocks, $row->text);       
	}
    
    protected function _place($blocks, &$text)
    {
        $text_length = strlen($text);
        if ($text_length <= 25000)
        {
            $max_n_blocks = intdiv($text_length, 5000);
        }
        else
        {
            $max_n_blocks = 5;
        }
        $n_blocks = min(count($blocks), $max_n_blocks);
        
        $n_paragraphs = substr_compare($text, '<p>');
        $start_offset = 5;
        $end_offset = 10;
        $n_paragraphs_avail = n_paragraphs - $start_offset - $end_offset;
        $blocks_freq = intdiv($n_paragraphs_avail, $n_blocks);
        
        $pos = strpos($text, '<p>');
        $count = 0;
        while ($pos !== false)
        {
            $offset = $pos + 3;
            $pos = strpos($text, '<p>', $offset)
            $count += 1;
            if ($count == $start_offset)
            {
                $block_id = array_shift($blocks);
                $block_id_str = "yandex_rtb_$block_id";
                substr_replace($text, "<div class=\"$block_id_str\"></div>\n<p>", $offset, 3);
                $start_offset += $blocks_freq;
            }
        }       
    }
}
?>
